### PR TITLE
Implemented the ticket transfer feature together with test cases.

### PR DESCRIPTION
**Pull request by bvanderdrift**

### CODE REVIEW

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+@RestController
+public class TicketTransferRestController {
+    @Autowired
+    private TicketService ticketService;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private TicketRepository ticketRepository;
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: With more then 2 @Autowired annotation, it's better to use a single @Autowired constructor

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    @PreAuthorize("isAuthenticated()")
+    @RequestMapping(value = "/tickettransfer", method = RequestMethod.POST)
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: I'd map this to /tickets/transfer

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public ResponseEntity<?> requestTicketTransfer(Authentication auth, @RequestBody @Validated TransferDTO transferDTO){
+        String username = transferDTO.getGoalUsername();
+        String ticketKey = transferDTO.getTicketKey();
+
+        User sender = (User)auth.getPrincipal();
+
+        User u = userService.getUserByUsername(username).orElseThrow(() -> new UsernameNotFoundException("User " + username + " not found."));
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: username can be inlined

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        Ticket t = ticketRepository.findByKey(ticketKey).orElseThrow(() -> new TicketNotFoundException(ticketKey));
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: ticketKey can be inlined

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        if(!t.getOwner().equals(sender)) return ResponseEntityBuilder.createResponseEntity(HttpStatus.FORBIDDEN, "You are not allowed to set this ticket up for transfer");
+
+        t.setUpForTransfer(u);
+
+        ticketRepository.save(t);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Move the logic to the TicketService

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+        return new ResponseEntity(HttpStatus.OK);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Use the `ch.wisv.areafiftylan.util.ResponseEntityBuilder.createResponseEntity()` function to create standard ResponseEntities

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+
+    @RequestMapping(value = "/tickettransfer/{ticketKey}", method = RequestMethod.PUT)
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: I'd map this to /tickets/transfer/{ticketKey}

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    public ResponseEntity<?> transferTicket(Authentication auth, @PathVariable String ticketKey){
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Authentication is null when the User is not logged in. Add a @PreAuthorize to make sure there's a user logged in

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        if (!t.isLockedForTransfer()){
+            throw new TicketNotTransferrableException(ticketKey);
+        }
+
+        if (!t.getTransferGoalOwner().equals(u)){
+            throw new NotGoalUserException();
+        }
+
+        ticketService.transferTicket(u, ticketKey);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Move all the logic to the TicketService, not just the transferTicket part

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        return new ResponseEntity(HttpStatus.OK);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Use the `ch.wisv.areafiftylan.util.ResponseEntityBuilder.createResponseEntity()` method to create a standard ResponseEntity

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Unneeded empty lines

File: src/main/java/ch/wisv/areafiftylan/dto/TransferDTO.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * Created by beer on 7-1-16.
+ */
+public class TransferDTO {
+
+    String ticketKey;
+
+    String goalUsername;
+
+    public String getTicketKey(){
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/dto/TransferDTO.java]: Replace the getters and setters with Lombok annotations, see the other DTO's how to do this.

File: src/main/java/ch/wisv/areafiftylan/model/Ticket.java [hunk 13]
```diff
@@ -63,11 +64,27 @@ public Ticket() {
         //JPA Only
     }
 
+    public void setUpForTransfer(User u){
+        this.setLockedForTransfer(true);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/model/Ticket.java]: You're locking the ticket for transfer if you want to set up a ticket for transfer?

File: src/main/java/ch/wisv/areafiftylan/model/Ticket.java [hunk 14]
```diff
@@ -63,11 +64,27 @@ public Ticket() {
...
+        this.setTransferGoalOwner(u);
+    }
+
+    public void finalizeTransfer(){
+        if(!this.isLockedForTransfer()) throw new TicketNotTransferrableException(this.key);
+
+        this.setLockedForTransfer(false);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/model/Ticket.java]: This is always false already, because else an exception would be thrown.

File: src/main/java/ch/wisv/areafiftylan/model/Ticket.java [hunk 15]
```diff
@@ -63,11 +64,27 @@ public Ticket() {
         //JPA Only
     }
 
+    public void setUpForTransfer(User u){
+        this.setLockedForTransfer(true);
+        this.setTransferGoalOwner(u);
+    }
+
+    public void finalizeTransfer(){
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/model/Ticket.java]: I'd move this logic to the service, as this is supposed to be just the model class

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+        receiver.getProfile()
+                .setAllFields("receiver", " of tickets", "GotYaTicket", Gender.MALE, "Money Owner 4", "2826GJ", "Tomorrowland",
+                        "0906-1111", null);
+
+        userRepository.saveAndFlush(receiver);
+
+        return receiver;
+    }
+
+    private Ticket makeMockTicket(){
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: This is not a mock ticket, it's an actual ticket ;)

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+        given()
+        .when()
+                .content(transferRequest).contentType(ContentType.JSON)
+                .post(TRANSFER_ENDPOINT)
+        .then()
+                .statusCode(HttpStatus.SC_FORBIDDEN);
+
+        ticket = ticketRepository.findByKey(ticket.getKey()).orElse(null);
+        if(ticket == null) Assert.fail("Could not refresh ticket");
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: Is this even possible? You just created and stored this ticket

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+        Assert.assertTrue(ticket.isLockedForTransfer());
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: If I understand correctly, this is only true of the ticket is ready for transfer. This is a fresh ticket, so it's not transferrable yet. Right?

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Test
+    public void testAddTransfer_Receiver(){
+        addTicketTransfer(ticketReciever.getUsername(), "password").then().statusCode(HttpStatus.SC_FORBIDDEN);
+
+        ticket = ticketRepository.findByKey(ticket.getKey()).orElse(null);
+        if(ticket == null) Assert.fail("Could not refresh ticket");
+
+        Assert.assertTrue(ticket.isLockedForTransfer());
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: Same here. The ticket was never prepared for transfer, then why should it be locked?

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Test
+    public void testAddTransfer_Owner(){
+        addTicketTransfer(user.getUsername(), "password").then().statusCode(HttpStatus.SC_OK);
+
+        ticket = ticketRepository.findByKey(ticket.getKey()).orElse(null);
+        if(ticket == null) Assert.fail("Could not refresh ticket");
+
+        Assert.assertTrue(ticket.isLockedForTransfer());
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: junit.Assert is deprecated. Please use a way that's not

**Comment by skamoen** [comment]
Please fix the formatting for the tests, it makes them more readable. See the other tests on how to do this.

**Comment by skamoen** [comment]
Should admins have the right to lock a ticket for transfer ans/or perform the transfer?

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    public ResponseEntity<?> requestTicketTransfer(Authentication auth, @RequestBody @Validated TransferDTO transferDTO){
+        String username = transferDTO.getGoalUsername();
+        String ticketKey = transferDTO.getTicketKey();
+
+        User sender = (User)auth.getPrincipal();
+
+        User u = userService.getUserByUsername(username).orElseThrow(() -> new UsernameNotFoundException("User " + username + " not found."));
+        Ticket t = ticketRepository.findByKey(ticketKey).orElseThrow(() -> new TicketNotFoundException(ticketKey));
+
+        if(!t.getOwner().equals(sender)) return createResponseEntity(HttpStatus.FORBIDDEN, "You are not allowed to set this ticket up for transfer");
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: Is it possible to solve this with a @PreAuthorize tag?

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+        ticketService.setupForTransfer(t, u);
+
+        return createResponseEntity(HttpStatus.OK, "Ticket succesfully set up for transfer");
+    }
+
+    @RequestMapping(value = "/tickets/transfer/{ticketKey}", method = RequestMethod.PUT)
+    public ResponseEntity<?> transferTicket(Authentication auth, @PathVariable String ticketKey){
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: auth can be null here if there's no user logged in. You can solve this with `@PreAuthorize("isAuthenticated()")` to enforce that someone is logged in before this function can be executed.

File: src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        User u = (User)auth.getPrincipal();
+
+        //Here an TicketNotTransferrable is thrown instead of ticket not found, otherwise there is an information leak since they know when a ticket doesn't exist or is just not transferrable
+        Ticket t = ticketRepository.findByKey(ticketKey).orElseThrow(() -> new TicketNotTransferrableException(ticketKey));
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/controller/TicketTransferRestController.java]: I'd move this to the transferTicket method, so you can remove the dependency on the TicketRepository completely from the Controller. Signature for would become something like `transferTicket(u, key)`

File: src/main/java/ch/wisv/areafiftylan/service/OrderServiceImpl.java [hunk 24]
```diff
@@ -36,10 +37,11 @@
 
     @Autowired
     public OrderServiceImpl(OrderRepository orderRepository, UserService userService, TicketRepository ticketRepository,
-                            PaymentService paymentService, ExpiredOrderRepository expiredOrderRepository) {
+                            TicketService ticketService, PaymentService paymentService, ExpiredOrderRepository expiredOrderRepository) {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/OrderServiceImpl.java]: It's weird that we use both the TicketRepository and the TicketService here. That's not part of this PR, but a reminder that I should fix this.

File: src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Optional;
+
+@Service
+public class TicketServiceImpl implements TicketService {
+
+    @Autowired
+    private TicketRepository ticketRepository;
+
+    @Override
+    public Optional<Ticket> getTicketByKey(String key){
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java]: Services should handle the Optional part, so that the Controller doesn't have to worry about nulls. You can simply throw the exception here.

File: src/main/java/ch/wisv/areafiftylan/service/TicketService.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @param type          Type of the Ticket requested
+     * @param owner         User that wants the Ticket
+     * @param pickupService If the Ticket includes the pickupService
+     *
+     * @return The requested ticket, if available
+     *
+     * @throws TicketUnavailableException If the requested ticket is sold out.
+     */
+    Ticket requestTicketOfType(TicketType type, User owner, boolean pickupService, boolean chMember);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/TicketService.java]: This is a duplicate of https://github.com/AreaFiftyLAN/lancie-api/blob/master/src/main/java/ch/wisv/areafiftylan/service/OrderService.java#L64?

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java (hunk 17 above)
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: Can you confirm that this is a useful Assert?

Comment (bvanderdrift) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: Well yeah, we should test if the ticket hasn't been put into transferrable state right? Testing that the api returns forbidden is not enough to test that the ticket in the backend hasn't been altered anyway right?

File: src/main/java/ch/wisv/areafiftylan/service/TicketService.java (hunk 26 above)
Comment (bvanderdrift) [inline src/main/java/ch/wisv/areafiftylan/service/TicketService.java]: Removed the linked duplicated. Seems more logical to be in the TicketService

**Comment by bvanderdrift** [comment]
@skamoen up for review again.

File: src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+        this.userService = userService;
+    }
+
+    @Override
+    public Ticket getTicketByKey(String key){
+        return ticketRepository.findByKey(key).orElseThrow(() -> new TicketNotFoundException("Ticket with key " + key + " not found."));
+    }
+
+    @Override
+    public Ticket requestTicketOfType(TicketType type, User owner, boolean pickupService, boolean chMember) {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java]: Can you make this a copy of the function you removed? I'm missing the `synchronized` here, not sure if there are other changes.

File: src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+        if (ticket.isTransferrable()) {
+            finalizeTransfer(ticket);
+
+            ticketRepository.save(ticket);
+        } else {
+            throw new TicketNotTransferrableException(ticket.getKey());
+        }
+    }
+
+    public void finalizeTransfer(Ticket t){
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java]: If this is not part of the interface, it should be private

File: src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+        if(!t.isTransferrable()) throw new TicketNotTransferrableException(t.getKey());
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java]: This is redundant, as you check it on line 56 as well.

File: src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Override
+    public void transferTicket(String ticketKey) {
+        Ticket ticket = ticketRepository.findByKey(ticketKey).orElseThrow(() -> new TicketNotFoundException(ticketKey));
+
+        if (ticket.isTransferrable()) {
+            finalizeTransfer(ticket);
+
+            ticketRepository.save(ticket);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java]: I would move this to the private function doing the transfer logic. If you want to reuse this function, it may get lost because it's not saved

Comment (bvanderdrift) [inline src/main/java/ch/wisv/areafiftylan/service/TicketServiceImpl.java]: Good call

**Comment by bvanderdrift** [comment]
back to @skamoen

**Comment by bvanderdrift** [comment]
Added cancel ticket functionality. Also improved testing of ticket transfer functionality. Back to @skamoen  for review

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 31]
```diff
@@ -0,0 +1,388 @@
...
+
+    private User makeTicketReceiver(){
+        User receiver = new User("receiver", new BCryptPasswordEncoder().encode("password"), "receiver@gmail.com");
+        receiver.getProfile()
+                .setAllFields("receiver", " of tickets", "GotYaTicket", Gender.MALE, "Money Owner 4", "2826GJ", "Tomorrowland",
+                        "0906-1111", null);
+
+        userRepository.saveAndFlush(receiver);
+
+        return receiver;
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: You should return the object returned by the saveAndFlush method here.

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 32]
```diff
@@ -0,0 +1,388 @@
...
+
+    private User makeOutsider(){
+        User outsider = new User("outsider", new BCryptPasswordEncoder().encode("password"), "outsider@gmail.com");
+        outsider.getProfile()
+                .setAllFields("Nottin", "Todoeo Witit", "Lookinin", Gender.FEMALE, "LoserStreet 1", "2826GJ", "China",
+                        "0906-3928", null);
+
+        userRepository.saveAndFlush(outsider);
+
+        return outsider;
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: Same here

**Comment by bvanderdrift** [comment]
Ticket transfers now work with tokens instead of logic integrated into the tickets themselves. Also users are now emailed once someone sends them a ticket.

File: src/main/java/ch/wisv/areafiftylan/dto/TransferDTO.java [hunk 33]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * Created by beer on 7-1-16.
+ */
+public class TransferDTO {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/dto/TransferDTO.java]: A DTO with just one field can be directly inlined in the Controller method. To do that, you can replace `@RequestBody @Validated TransferDTO transferDTO` with `@RequestBody String goalUsername`. Does essentially the same thing.

File: src/main/java/ch/wisv/areafiftylan/exception/InvalidTokenException.java [hunk 34]
```diff
@@ -0,0 +1,15 @@
+package ch.wisv.areafiftylan.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+/**
+ * Created by beer on 11-3-16.
+ */
+
+@ResponseStatus(HttpStatus.UNAUTHORIZED)
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/exception/InvalidTokenException.java]: I'd turn this in something else then UNAUTHORIZED, because that's not the case. Maybe BAD_REQUEST or NOT_MODIFIED.

File: src/main/java/ch/wisv/areafiftylan/exception/NotGoalUserException.java [hunk 35]
```diff
@@ -0,0 +1,11 @@
+package ch.wisv.areafiftylan.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.UNAUTHORIZED)
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/exception/NotGoalUserException.java]: FORBIDDEN is a better status for this exception

File: src/main/java/ch/wisv/areafiftylan/exception/TicketNotTransferrableException.java [hunk 36]
```diff
@@ -0,0 +1,12 @@
+package ch.wisv.areafiftylan.exception;
+
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+public class TicketNotTransferrableException extends RuntimeException{
+    public TicketNotTransferrableException(String ticketKey) {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/exception/TicketNotTransferrableException.java]: Is this still used?

File: src/main/java/ch/wisv/areafiftylan/model/util/TransferState.java [hunk 37]
```diff
@@ -0,0 +1,8 @@
+package ch.wisv.areafiftylan.model.util;
+
+/**
+ * Created by beer on 3-1-16.
+ */
+public enum TransferState {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/model/util/TransferState.java]: This is not used.

File: src/main/java/ch/wisv/areafiftylan/security/TicketTransferToken.java [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+@Entity
+public class TicketTransferToken extends Token {
+
+    @OneToOne(targetEntity = Ticket.class, fetch = FetchType.EAGER)
+    @JoinColumn(nullable = false)
+    private Ticket ticket;
+
+    @OneToOne(targetEntity = User.class, fetch = FetchType.EAGER)
+    @JoinColumn(nullable = false)
+    private User goalUser;
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/security/TicketTransferToken.java]: You can remove this and just use the `Token` user as goal user. We don't need to store the Ticket sender, as this is the Ticket owner

File: src/main/java/ch/wisv/areafiftylan/security/Token.java [hunk 39]
```diff
@@ -66,13 +68,21 @@ public void setExpiryDate(Date expiryDate) {
         this.expiryDate = expiryDate;
     }
 
-    public void setUsed(boolean used) {
-        this.used = used;
+    public boolean getIsExpireable(){
+        return EXPIRATION != 0;
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/security/Token.java]: This is currently not used in the `isExpired()` method, so it's not working as intended. the `isExpired()` is part of `isValid()` so Tokens with `EXPIRATION = 0` no get expired immediately.

File: src/main/java/ch/wisv/areafiftylan/service/CurrentUserService.java [hunk 40]
```diff
@@ -1,5 +1,7 @@
 package ch.wisv.areafiftylan.service;
 
+import java.util.Objects;
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/CurrentUserService.java]: Unused import

File: src/main/java/ch/wisv/areafiftylan/service/MailServiceImpl.java [hunk 41]
```diff
@@ -92,4 +92,10 @@ public void sendPasswordResetMail(User user, String url) throws MessagingExcepti
         String message = "Please click on the following link to reset your password: " + url;
         sendMail(user.getEmail(), user.getUsername(), null, "Password reset requested", message);
     }
+
+    @Override
+    public void sendTicketTransferMail(User sender, User receiver, String url) throws MessagingException {
+        String message = "The user \"" + sender.getUsername() + "\" has send you a ticket for AreaFiftyLAN! To accept this ticket please click on the following link: " + url;
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/service/MailServiceImpl.java]: I'd change this to include the full name of the sender. Tickets were bought, so a filled in profile is guaranteed. The term "user" isn't really something you want to say. They're people! :stuck_out_tongue_winking_eye:

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 42]
```diff
@@ -0,0 +1,335 @@
...
+    private User ticketReceiver;
+    private Ticket ticket;
+
+    @Autowired
+    private TicketRepository ticketRepository;
+
+    @Autowired
+    private TicketTransferTokenRepository tttRepository;
+
+
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: Blank line

File: src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java [hunk 43]
```diff
@@ -0,0 +1,335 @@
...
+    }
+
+    @Test
+    public void testAddTransfer_Anon(){
+        Map<String, String> transferRequest = new HashMap<>();
+        transferRequest.put("goalUsername", ticketReceiver.getUsername());
+
+        given().
+        when().
+                content(transferRequest + "/" + ticket.getId()).contentType(ContentType.JSON).
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/TicketTransferRestIntegrationTest.java]: This is weird. Test is wrong, but the outcome is correct anyway because there's not CSRF token

```
{
    "timestamp": 1457724563201,
    "status": 403,
    "error": "Forbidden",
    "message": "Expected CSRF token not found. Has your session expired?",
    "path": "/tickets/transfer"
}
```

File: src/main/java/ch/wisv/areafiftylan/security/TicketTransferToken.java (hunk 38 above)
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/security/TicketTransferToken.java]: Changed my mind. As Admins can also invite people, it might be useful to store the sender anyway.

**Comment by bvanderdrift** [comment]
Aah damn. I'll jump on it next monday.

File: src/main/java/ch/wisv/areafiftylan/service/OrderServiceImpl.java (hunk 24 above)
Comment (bvanderdrift) [inline src/main/java/ch/wisv/areafiftylan/service/OrderServiceImpl.java]: Have you made a ticket of this yet for yourself?

File: src/main/java/ch/wisv/areafiftylan/dto/TransferDTO.java (hunk 33 above)
Comment (bvanderdrift) [inline src/main/java/ch/wisv/areafiftylan/dto/TransferDTO.java]: On a side note: It doesn't do the same thing. If replaced with string the entire JSON string will be injected into the goalUsername field and not just the corresponding field of the JSON object.

**Comment by bvanderdrift** [comment]
Back to @skamoen

**Comment by skamoen** [comment]
LGTM. I fixed unused imports and extracted the ticketRepository, can be merged when Travis is happy
